### DISCUSSION TITLE
Branching Strategies

### DISCUSSION THREAD

**Discussion by graciegoheen** [dbt-labs/docs.getdbt.com#1315] [Archive]
**1. What is the main problem you are solving? What is your solution? This should help form your core thesis.**
   What is our recommended branching strategy for customers who are using a multi-branch approach? (Example: Dev -> UAT -> Prod -> Master)
   
**2. Why should the reader care about this problem? Why is your solution the right one? This should help form your specific target audience.**
  Introducing more branches adds lots of complexity! It can be challenging to understand the pros and cons and decide what strategy is best for your team.

**3. Can you list the steps of your solution for the reader here? This should help you form the overall narrative arc and sketch out an example use case to illustrate it.**
  First, it's important to consider _what_ you are trying to accomplish by having multiple-branches, and why a simple workflow (Example: just one Main branch) doesn't solve it. 

   If your business _requires_ you have a separate `Dev`, `UAT`, `Prod`, and `Master` branch, you should follow this approach:
  ![image](https://user-images.githubusercontent.com/53586774/162281310-b20ab094-035e-4a14-9cf6-297758fefdb5.png)
 
- checkout `dev` when creating your Feature Branches
- merge the Feature Branches back into `dev` once they get the stamp of approval
- `dev` gets merged into `UAT`, `UAT` to `prod`, and finally `prod` to `master`
- if issues / bugs pop up during testing, create hot-fixes as necessary until the *entire* `UAT` branch is ready to be merged into `Prod`

More thoughts [here](https://www.notion.so/dbtlabs/Our-Thoughts-Branching-Strategies-1cf59cba4a6c430596c0a12c07af5b52).

**4. Are there any resources that helped inspire or inform your idea? (eg slack discussions, articles, external product docs, etc. -- if so please link)**
   [Adopting CI/CD with dbt Cloud](https://blog.getdbt.com/adopting-ci-cd-with-dbt-cloud/)

**Comment by gwenwindflower** [comment]
yesss i love this! we absolutely need a best practices guide on branching strategies, and this is also a great way to bridge down the road into how these things intersect with CI/CD best practices. perfect guide material! 

i do think this is a pretty big + important topic though so it's something we should get a couple more proserv and/or solutions architects to weigh in on, or just working AE perspectives generally -- couple ways we could do that:

1. keep this tight and just tag a few people specifically to jump in and synthesize this approach with their own to broaden/strengthen it/verify it
2. open a fresh `I want to learn about...` Discussion and try drive a discussion with the Community at large in addition to dbt Labbers

either way is good! i'll offer some questions and thoughts for the knowledge directly in a follow-up comment shortly.

**Comment by graciegoheen** [comment]
Opened up an issue! https://github.com/dbt-labs/docs.getdbt.com/issues/1704
